java-checks/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java

```diff
@@ -187,17 +231,215 @@ public void visitVariable(VariableTree tree) {
+    lastSyntaxToken = tree.identifierToken();
+  }
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    if (!tree.types().isEmpty()) {
+      scan(Iterables.getLast(tree.types()));
+    } else if (!tree.imports().isEmpty()) {
+      scan(Iterables.getLast(tree.imports()));
+    } else {
+      // TODO(SONARJAVA-547) Should be the semi-colon token
+      scan(tree.packageName());
+    }
+    // with empty files lastSyntaxToken will be null
+  }
+
+  @Override
+  public void visitImport(ImportTree tree) {
+    lastSyntaxToken = tree.semicolonToken();
+  }
+
+  @Override
+  public void visitLabeledStatement(LabeledStatementTree tree) {
+    scan(tree.statement());
+  }
+
+  @Override
+  public void visitCaseGroup(CaseGroupTree tree) {
+    if (!tree.body().isEmpty()) {
+      scan(Iterables.getLast(tree.body()));
+    } else {
+      scan(Iterables.getLast(tree.labels()));
+    }
+  }
+
+  @Override
+  public void visitUnaryExpression(UnaryExpressionTree tree) {
+    if (tree.is(Tree.Kind.POSTFIX_DECREMENT, Tree.Kind.POSTFIX_INCREMENT)) {
+      lastSyntaxToken = tree.operatorToken();
+    } else {
+      scan(tree.expression());
+    }
+  }
+
+  @Override
+  public void visitConditionalExpression(ConditionalExpressionTree tree) {
+    scan(tree.falseExpression());
+  }
+
+  @Override
+  public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+    lastSyntaxToken = tree.closeBracketToken();
+  }
+
+  @Override
+  public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+    scan(tree.identifier());
+  }
+
+  @Override
+  public void visitNewClass(NewClassTree tree) {
+    if (tree.classBody() != null) {
+      scan(tree.classBody());
+    } else if (tree.closeParenToken() != null) {
+      lastSyntaxToken = tree.closeParenToken();
+    } else {
+      scan(tree.identifier());
+    }
+  }
+
+  @Override
+  public void visitNewArray(NewArrayTree tree) {
+    if (!tree.initializers().isEmpty()) {
+      // FIXME should be the close brace
+      scan(Iterables.getLast(tree.initializers()));
+    } else {
+      scan(Iterables.getLast(tree.dimensions()));
+    }
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    lastSyntaxToken = tree.closeParenToken();
+  }
+
+  @Override
+  public void visitTypeCast(TypeCastTree tree) {
+    scan(tree.expression());
+  }
+
+  @Override
+  public void visitInstanceOf(InstanceOfTree tree) {
+    scan(tree.type());
+  }
+
+  @Override
+  public void visitParenthesized(ParenthesizedTree tree) {
+    lastSyntaxToken = tree.closeParenToken();
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    scan(tree.expression());
+  }
+
+  @Override
+  public void visitPrimitiveType(PrimitiveTypeTree tree) {
+    lastSyntaxToken = tree.keyword();
+  }
+
+  @Override
+  public void visitArrayType(ArrayTypeTree tree) {
+    // FIXME should be the last dimension
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)